### DISCUSSION TITLE
Dask doesn't reload code

### DISCUSSION THREAD

**Discussion by jadeidev** [dask/dask#8514] [Q&A] [answered]
I started a local dask cluster 
```console
$ dask-scheduler
Scheduler started at 127.0.0.1:8786

$ dask-worker 127.0.0.1:8786
$ dask-worker 127.0.0.1:8786
```
 and ran my project and I have noticed that if I change code in one of my project submodules after running for the first time dask would still run the old code. is this the intended behavior? Admittedly the example below is a little lengthy so I apologize but it will demonstrate the point. 
Consider the following small project 
#### project tree
```
~/test
│   example.log
│   setup.py
└───src
    └──myproject
           dask_toy.py
           main.py
           __init__.py
```
#### setup.py
```python
from setuptools import setup, find_packages

setup(
    name="myproject",
    version="1.0",
    packages=find_packages(where="src"),
    package_dir={"": "src"},
    entry_points={"console_scripts": ["myproject = myproject.main:main"]},
)
```
#### main.py
```python
from dask.distributed import Client
import dask
from myproject.dask_toy import waiting_method


def main():
    client = Client("127.0.0.1:8786")  # change this after launching the scheduler
    # client.restart()

    test = [dask.delayed(waiting_method)(1), dask.delayed(waiting_method)(2)]
    dask.compute(*test)


if __name__ == "__main__":
    main()
```

#### dask_toy.py
```python
from time import sleep
import logging
logging.basicConfig(format='%(asctime)s %(message)s',filename='example.log', filemode='w', level=logging.INFO)


def waiting_method(x):
    logging.info(f"start sleep {x}")
    sleep(5)
    logging.info(f"stop sleep {x}")
    return "SUCCESS"
```

#### console:
**start python virtual env and activate it**
```console
$ cd ~/test
$ pip install dask[distributed]
$ pip install -e .
```
start dask scheduler and 2 workers within the python environment in 3 different consoles and update `main.py` with the correct dask scheduler `"x.y.z.p:port"` and  save and run `$ myproject` . 
Now while dask scheduler and workers are still running change the sleep time on `dask_toy.py` to 15 like so:
```python
def waiting_method(x):
    logging.info(f"start sleep {x}")
    sleep(15) #<------------------------- change to 15
    logging.info(f"stop sleep {x}")
    return "SUCCESS"
```
 save and and rerun: `$ myproject `

### result on example.log:
```
2021-12-22 14:48:06,238 start sleep 2
2021-12-22 14:48:06,238 start sleep 1
2021-12-22 14:48:11,244 stop sleep 1
2021-12-22 14:48:11,244 stop sleep 2
after change to 15 seconds in dask_toy.py the delta between start and stop remains 5 seconds
2021-12-22 14:51:23,860 start sleep 1
2021-12-22 14:51:23,860 start sleep 2
2021-12-22 14:51:28,875 stop sleep 1
2021-12-22 14:51:28,876 stop sleep 2
```
if we want to update the code, we need a hard restart in `main.py`. after that the time between start and stop will indeed change to 15 sec. like so:
```python
from dask.distributed import Client
import dask
from myproject.dask_toy import waiting_method


def main():
    client = Client("127.0.0.1:8786")  # change this after launching the scheduler
    client.restart()         #<-------------------- ADD THIS LINE

    test = [dask.delayed(waiting_method)(1), dask.delayed(waiting_method)(2)]
    dask.compute(*test)
```

**Comment by SultanOrazbayev** [comment]
Not sure if this is helpful, but have you considered [`importlib.reload`](https://stackoverflow.com/a/33405315/10693596)?

**Reply by abergou** [reply]
You may also want to try [xreload](https://deshaw.github.io/pyflyby/api/livepatch.html?highlight=reload) from the [pyflyby](https://github.com/deshaw/pyflyby) python module. Although there are various caveats, you could implement a worker plugin that periodically runs xreload() on a worker.

**Comment by jadeidev** [comment, marked as answer]
See issue #8516
